src/System.Text.CaseFolding/System/Text/CaseFolding/SimpleCaseFolding.cs

```diff
@@ -0,0 +1,630 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.CaseFolding
+{
+    /// <summary>
+    /// Implement Unicode Simple Case Folding API.
+    /// </summary>
+    public static partial class SimpleCaseFolding
+    {
+        private static ref ushort s_MapLevel1 => ref MapLevel1[0];
+        private static ref char s_refMapData => ref MapData[0];
+        private static ref ushort s_refMapSurrogateLevel1 => ref MapSurrogateLevel1[0];
+        private static ref (char highSurrogate, char lowSurrogate) s_refMapSurrogateData => ref MapSurrogateData[0];
+
+        /// <summary>
+        /// Simple case fold the char.
+        /// </summary>
+        /// <param name="c">Source char.</param>
+        /// <returns>
+        /// Returns folded char.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static char SimpleCaseFold(char c)
+        {
+            if (c <= 0x5ff)
+            {
+                return MapBelow5FF[c];
```

Comment:
BenchmarkDotNet shows that your pattern does the boundary check twice and original code - only one. Perhaps it is a bug in CoreCLR.

Reply:
That is something we should track with a bug in coreclr. Can you please share the disassembly as well for before/after?

cc @AndyAyersMS

Reply:
First question is - should we think about 2.1 or only about 3.0? I tried to make the code the fastest on 2.1. If we target only to 3.0, then all code can be changed in some places.

Reply:
We should write the code to work best for .NET Core 3.0 and measure perf against that (and not .NET Core 2.1), especially if the end goal of this API is to ship as part of .NET Core vNext. There have been quite a few changes in the JIT/CoreCLR between 2.1 and 3.0.

Reply:
Thanks! Can we postpone optimizations to follow PR where I'd review all code and report all perf issues if found (the PR already has huge connents)?

Reply:
@ahsonkhan I tested your sample (with SimpleCaseFoldTest name). 
? - it is 0x600 char in output.
 (I can not attach file with results because of GitHub error.) 

SimpleCaseFold(char) win in all tests besides test (1.7495 ns) before last where optimization is broken.
If we want SimpleCaseFoldTest(char) to win we need better runtime optimizations.

In addition, I would say that there can be probably made several optimizations for static readonly arrays
- reduce prolog code to access  static readonly arrays (?)
- remove boundary checks for first item - MapBelow5FF[0]

Could you initiate review from appropriate Core team? 

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview-010190
  [Host]      : .NET Core 3.0.0-preview-27329-4 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT
  TC Disabled : .NET Core 3.0.0-preview-27329-4 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT
  TC Enabled  : .NET Core 3.0.0-preview-27329-4 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT

EnvironmentVariables=COMPlus_TieredCompilation=1  Runtime=Core  Toolchain=.NET Core 3.0  

```
|             Method |         Job | TestChar |      Mean |     Error |    StdDev |    Median | Ratio | RatioSD |
|------------------- |------------ |--------- |----------:|----------:|----------:|----------:|------:|--------:|
|     **SimpleCaseFold** | **TC Disabled** |        **A** | **0.4927 ns** | **0.0542 ns** | **0.0453 ns** | **0.4750 ns** |  **1.00** |    **0.00** |
| SimpleCaseFoldTest | TC Disabled |        A | 1.1295 ns | 0.0498 ns | 0.0416 ns | 1.1149 ns |  2.31 |    0.22 |
|                    |             |          |           |           |           |           |       |         |
|     SimpleCaseFold |  TC Enabled |        A | 0.7198 ns | 0.0478 ns | 0.0424 ns | 0.7006 ns |  1.00 |    0.00 |
| SimpleCaseFoldTest |  TC Enabled |        A | 1.0581 ns | 0.0195 ns | 0.0163 ns | 1.0582 ns |  1.47 |    0.09 |
|                    |             |          |           |           |           |           |       |         |
|     **SimpleCaseFold** | **TC Disabled** |        **؀** | **0.5648 ns** | **0.3442 ns** | **0.7769 ns** | **0.1235 ns** |  **1.00** |    **0.00** |
| SimpleCaseFoldTest | TC Disabled |        ؀ | 1.9968 ns | 0.0411 ns | 0.0343 ns | 1.9972 ns |  1.05 |    0.08 |
|                    |             |          |           |           |           |           |       |         |
|     SimpleCaseFold |  TC Enabled |        ؀ | 1.7495 ns | 0.0238 ns | 0.0199 ns | 1.7532 ns |  1.00 |    0.00 |
| SimpleCaseFoldTest |  TC Enabled |        ؀ | 0.0852 ns | 0.0729 ns | 0.1314 ns | 0.0346 ns |  0.06 |    0.10 |

Reply:
Can you please share your benchmark code? Create a gist if that works better: https://gist.github.com/

Reply:
I tested 
```c#
if (c <= MapBelow5FF.Length)
{
   return  MapBelow5FF[c];
}
```
Code is in my repo:
https://github.com/iSazonov/SCF/blob/master/src/SimpleCaseFolding.cs#L53-L56
https://github.com/iSazonov/SCF/blob/master/tests/Benchmarks/StringFolding/Perf.StringFolding.cs